File: Trie/4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        
        
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1
        n1 = len(nums1)
        n2 = len(nums2)
        total = n1+n2 
        half = (total) // 2


        l = 0
        r = n1 - 1
        while True:
            i = (l+r)//2
            j = half - i - 2

            aLeft = nums1[i] if i >= 0 else float('-inf')
            aRight = nums1[i+1] if i+1 < n1 else float('inf')
            bLeft = nums2[j] if j >= 0 else float('-inf')
            bRight = nums2[j+1] if j+1 < n2 else float('inf')

            if aLeft <= bRight and bLeft <= aRight:
                if total % 2:
                    return min(aRight, bRight)
                return (min(aRight, bRight) + max(aLeft , bLeft)) / 2
            elif aLeft > bRight:
                r = i-1
            else:
                l = i+1

File: Trie/4-median-of-two-sorted-arrays/median-of-two-sorted-arrays.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function(nums1, nums2) {
    
    if (nums1.length > nums2.length){
        [nums1, nums2] = [nums2, nums1]
    }

    const n1 = nums1.length
    const n2 = nums2.length

    total = n1+n2
    half = Math.floor(total/2)

    l = 0
    r = n1-1
    while(true) {
        n1Mid = Math.floor((l+r)/2)
        n2Mid = half - n1Mid - 2
        aLeft = n1Mid < 0 ? -Infinity : nums1[n1Mid]
        aRight = n1Mid + 1 >= n1 ? Infinity: nums1[n1Mid + 1]
        bLeft = n2Mid < 0 ? -Infinity : nums2[n2Mid]
        bRight = n2Mid + 1 >= n2 ? Infinity: nums2[n2Mid + 1]

        if(aLeft <= bRight && bLeft <= aRight) {
            if(total%2 === 0) {
                return (Math.max(aLeft, bLeft) + Math.min(aRight, bRight))/2
            } else{
                return Math.min(aRight, bRight)
            }
        } else if (aLeft > bRight) {
            r = n1Mid-1
        } else {
            l = n1Mid+1
        }
    }


};